File: packages/skill-eval-python/src/skill_eval/runner/orchestrator.py
from __future__ import annotations

import asyncio
import tempfile
from collections.abc import Callable
from dataclasses import dataclass

from skill_eval.config.schema import ResolvedConfig
from skill_eval.grader.grader import grade_run
from skill_eval.providers.factory import create_provider
from skill_eval.providers.pricing import estimate_cost
from skill_eval.providers.types import LLMProvider
from skill_eval.schemas.benchmark import BenchmarkRun, RunResult
from skill_eval.schemas.evals import EvalCase, get_turns
from skill_eval.schemas.grading import ExpectationResult, GradingResult
from skill_eval.utils.async_utils import with_timeout
from skill_eval.utils.error import extract_error_message
from skill_eval.utils.math_utils import mean, stddev
from skill_eval.utils.rate import rate_color

from .agent_loop import run_agent_loop
from .sandbox import collect_output_files, seed_sandbox
from .tools import DEFAULT_TOOLS, create_tool_handlers
from .turn_check import create_turn_checker

try:
    from rich import print as rprint
except ImportError:
    rprint = print

MAX_RETRIES = 2
RETRY_BASE_MS = 2000
MAX_CONCURRENCY = 10

RETRYABLE_STATUS_CODES = {429, 502, 503, 529}
RETRYABLE_MESSAGES = ["rate limit", "overloaded", "unavailable", "high demand"]


@dataclass
class RunEntry:
    eval_case: EvalCase
    provider: LLMProvider
    run_number: int


def _is_retryable(err: BaseException) -> bool:
    e = getattr(err, "__dict__", {}) or {}
    status = e.get("status") or e.get("code")
    if isinstance(status, int) and status in RETRYABLE_STATUS_CODES:
        return True
    msg = str(getattr(err, "message", str(err))).lower()
    return any(m in msg for m in RETRYABLE_MESSAGES)


def _task_id(entry: RunEntry) -> str:
    return f"{entry.eval_case.id}-{entry.provider.model_id}-{entry.run_number}"


def _task_label(entry: RunEntry) -> str:
    return f"eval {entry.eval_case.id} · {entry.provider.model_id} · r{entry.run_number}"


def _build_failure_result(entry: RunEntry, err_msg: str) -> BenchmarkRun:
    return BenchmarkRun(
        eval_id=entry.eval_case.id,
        provider=entry.provider.name,
        model=entry.provider.model_id,
        run_number=entry.run_number,
        result=RunResult(
            pass_rate=0,
            passed=0,
            failed=len(entry.eval_case.assertions),
            total=len(entry.eval_case.assertions),
            time_seconds=0,
            input_tokens=0,
            output_tokens=0,
            total_tokens=0,
            tool_calls=0,
            errors=1,
            cost_usd=0,
        ),
        expectations=[
            ExpectationResult(text=a, passed=False, evidence=f"Error: {err_msg}")
            for a in entry.eval_case.assertions
        ],
        claims=[],
        eval_feedback=None,
        error=err_msg,
    )


def _compute_provider_summary(
    runs: list[BenchmarkRun],
) -> dict[str, dict[str, dict[str, float]]]:
    grouped: dict[str, list[BenchmarkRun]] = {}
    for run in runs:
        key = run.model
        if key not in grouped:
            grouped[key] = []
        grouped[key].append(run)

    summary: dict[str, dict[str, dict[str, float]]] = {}
    for key, model_runs in grouped.items():
        def _stat(fn, runs=model_runs):
            vals = [fn(r) for r in runs]
            return {"mean": mean(vals), "stddev": stddev(vals)}

        summary[key] = {
            "pass_rate": _stat(lambda r: r.result.pass_rate),
            "time_seconds": _stat(lambda r: r.result.time_seconds),
            "total_tokens": _stat(lambda r: r.result.total_tokens),
            "cost_usd": _stat(lambda r: r.result.cost_usd),
        }
    return summary


def _format_success_line(
    completed: int, total: int, entry: RunEntry, result: BenchmarkRun
) -> str:
    r = result.result
    color_fn = rate_color(r.pass_rate)
    return (
        f"  [green]✓[/green] [{completed}/{total}] "
        f"eval {entry.eval_case.id} · {entry.provider.model_id} · "
        f"run {entry.run_number} · "
        f"{color_fn(f'{r.pass_rate * 100:.0f}%')} [dim]({r.passed}/{r.total})[/dim] · "
        f"[dim]{r.time_seconds:.1f}s · {r.total_tokens} tok · "
        f"${r.cost_usd:.4f} · {r.tool_calls} calls[/dim]"
    )


async def run_orchestrator(
    config: ResolvedConfig,
    evals: list[EvalCase],
    skill_dir: str,
    system_prompt: str,
    *,
    concurrency: int | None = None,
    on_progress: Callable[[str], None] | None = None,
) -> dict:
    log = on_progress or (lambda msg: rprint(msg))

    grader_provider = create_provider(config.grader)

    matrix: list[RunEntry] = []
    for eval_case in evals:
        for provider_config in config.providers:
            provider = create_provider(provider_config)
            for run in range(1, config.settings.runs_per_provider + 1):
                matrix.append(
                    RunEntry(eval_case=eval_case, provider=provider, run_number=run)
                )

    concurrency_limit = concurrency or min(len(matrix), MAX_CONCURRENCY)

    log(f"[bold]Running {len(matrix)} eval(s): {len(evals)} eval(s) × "
        f"{len(config.providers)} provider(s) × {config.settings.runs_per_provider} run(s)[/bold]")

    results: list[BenchmarkRun] = []
    from .spinner import Spinner

    spinner = Spinner()
    completed = 0
    queue: asyncio.Queue[RunEntry | None] = asyncio.Queue()
    for entry in matrix:
        await queue.put(entry)
    for _ in range(concurrency_limit):
        await queue.put(None)

    async def execute_one(entry: RunEntry) -> BenchmarkRun:
        eval_timeout_sec = config.settings.timeout
        sandbox_dir = tempfile.mkdtemp(prefix="skill-eval-")
        try:
            seed_sandbox(sandbox_dir, skill_dir, entry.eval_case.files)
            turns = get_turns(entry.eval_case)

            agent_run = await with_timeout(
                run_agent_loop(
                    provider=entry.provider,
                    system=system_prompt,
                    turns=turns,
                    tools=DEFAULT_TOOLS,
                    tool_handlers=create_tool_handlers(
                        sandbox_dir, int(config.settings.timeout)
                    ),
                    max_steps=config.settings.max_steps,
                    temperature=config.settings.temperature,
                    on_activity=lambda d: spinner.detail(_task_id(entry), d),
                    check_turn_relevance=(
                        create_turn_checker(grader_provider)
                        if len(turns) > 1
                        else None
                    ),
                ),
                eval_timeout_sec,
                f"Eval timed out after {config.settings.timeout}s",
            )

            output_files = collect_output_files(sandbox_dir)
            spinner.detail(_task_id(entry), "grading…")
            grading: GradingResult = await grade_run(
                grader_provider, entry.eval_case, agent_run, output_files
            )

            cost = estimate_cost(
                entry.provider.model_id,
                agent_run.total_input_tokens,
                agent_run.total_output_tokens,
            )

            return BenchmarkRun(
                eval_id=entry.eval_case.id,
                provider=entry.provider.name,
                model=entry.provider.model_id,
                run_number=entry.run_number,
                result=RunResult(
                    pass_rate=grading.pass_rate,
                    passed=grading.passed,
                    failed=grading.failed,
                    total=grading.total,
                    time_seconds=agent_run.total_latency_ms / 1000,
                    input_tokens=agent_run.total_input_tokens,
                    output_tokens=agent_run.total_output_tokens,
                    total_tokens=agent_run.total_input_tokens
                    + agent_run.total_output_tokens,
                    tool_calls=agent_run.total_tool_calls,
                    errors=agent_run.errors,
                    cost_usd=cost,
                ),
                expectations=grading.expectations,
                claims=grading.claims,
                eval_feedback=grading.eval_feedback,
                error=None,
            )
        finally:
            import shutil

            shutil.rmtree(sandbox_dir, ignore_errors=True)

    async def run_task(entry: RunEntry) -> None:
        nonlocal completed
        tid = _task_id(entry)
        spinner.track(tid, _task_label(entry))

        attempt = 0
        while True:
            try:
                result = await execute_one(entry)
                results.append(result)
                completed += 1
                spinner.succeed(
                    tid,
                    _format_success_line(completed, len(matrix), entry, result),
                )
                return
            except Exception as err:
                attempt += 1
                err_msg = extract_error_message(err)

                if _is_retryable(err) and attempt <= MAX_RETRIES:
                    delay_ms = RETRY_BASE_MS * (2 ** (attempt - 1))
                    spinner.detail(
                        tid,
                        f"retrying in {delay_ms / 1000:.0f}s "
                        f"({attempt + 1}/{MAX_RETRIES + 1})",
                    )
                    await asyncio.sleep(delay_ms / 1000)
                    spinner.untrack(tid)
                    spinner.track(tid, _task_label(entry))
                    continue

                results.append(_build_failure_result(entry, err_msg))
                completed += 1
                spinner.succeed(
                    tid,
                    f"  [red]✗[/red] [{completed}/{len(matrix)}] "
                    f"eval {entry.eval_case.id} · {entry.provider.model_id} · "
                    f"run {entry.run_number} · [red]{err_msg}[/red]",
                )
                return

    async def worker() -> None:
        while True:
            entry = await queue.get()
            if entry is None:
                break
            await run_task(entry)

    spinner.start(len(matrix))
    await asyncio.gather(*[worker() for _ in range(concurrency_limit)])
    spinner.stop()

    return {
        "runs": results,
        "providerSummary": _compute_provider_summary(results),
    }

File: packages/skill-eval/src/runner/orchestrator.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import pc from "picocolors";
import type { ResolvedConfig } from "../config/schema.js";
import { gradeRun } from "../grader/grader.js";
import { createProvider } from "../providers/factory.js";
import { estimateCost } from "../providers/pricing.js";
import type { LLMProvider } from "../providers/types.js";
import type { BenchmarkRun } from "../schemas/benchmark.js";
import type { EvalCase } from "../schemas/evals.js";
import type { GradingResult } from "../schemas/grading.js";
import { withTimeout } from "../utils/async.js";
import { extractErrorMessage } from "../utils/error.js";
import { mean, sleep, stddev } from "../utils/math.js";
import { rateColor } from "../utils/rate.js";
import { runAgentLoop } from "./agent-loop.js";
import { Spinner } from "./spinner.js";
import { collectOutputFiles, createToolHandlers, defaultTools, seedSandbox } from "./tools.js";
import { createTurnChecker } from "./turn-check.js";

export interface RunEntry {
	evalCase: EvalCase;
	provider: LLMProvider;
	runNumber: number;
}

export interface OrchestratorResult {
	runs: BenchmarkRun[];
	providerSummary: Record<
		string,
		{
			pass_rate: { mean: number; stddev: number };
			time_seconds: { mean: number; stddev: number };
			total_tokens: { mean: number; stddev: number };
			cost_usd: { mean: number; stddev: number };
		}
	>;
}

const MAX_RETRIES = 2;
const RETRY_BASE_MS = 2000;
const MAX_CONCURRENCY = 10;

const RETRYABLE_STATUS_CODES = new Set([429, 502, 503, 529]);
const RETRYABLE_MESSAGES = ["rate limit", "overloaded", "unavailable", "high demand"];

function isRetryable(err: unknown): boolean {
	const e = err as Record<string, unknown> | undefined;
	const status = (e?.status ?? e?.code ?? (e?.error as Record<string, unknown>)?.code) as
		| number
		| undefined;
	if (typeof status === "number" && RETRYABLE_STATUS_CODES.has(status)) return true;
	const msg = String((err as Error)?.message ?? "").toLowerCase();
	return RETRYABLE_MESSAGES.some((m) => msg.includes(m));
}

function taskId(entry: RunEntry): string {
	return `${entry.evalCase.id}-${entry.provider.modelId}-${entry.runNumber}`;
}

function taskLabel(entry: RunEntry): string {
	return `eval ${entry.evalCase.id} · ${entry.provider.modelId} · r${entry.runNumber}`;
}

function buildFailureResult(entry: RunEntry, errMsg: string): BenchmarkRun {
	return {
		eval_id: entry.evalCase.id,
		provider: entry.provider.name,
		model: entry.provider.modelId,
		run_number: entry.runNumber,
		result: {
			pass_rate: 0,
			passed: 0,
			failed: entry.evalCase.assertions.length,
			total: entry.evalCase.assertions.length,
			time_seconds: 0,
			input_tokens: 0,
			output_tokens: 0,
			total_tokens: 0,
			tool_calls: 0,
			errors: 1,
			cost_usd: 0,
		},
		expectations: entry.evalCase.assertions.map((a) => ({
			text: a,
			passed: false,
			evidence: `Error: ${errMsg}`,
		})),
		claims: [],
		eval_feedback: null,
		error: errMsg,
	};
}

function computeProviderSummary(runs: BenchmarkRun[]): OrchestratorResult["providerSummary"] {
	const grouped = new Map<string, BenchmarkRun[]>();
	for (const run of runs) {
		const key = run.model;
		if (!grouped.has(key)) grouped.set(key, []);
		grouped.get(key)?.push(run);
	}

	const summary: OrchestratorResult["providerSummary"] = {};
	for (const [key, modelRuns] of grouped) {
		const stat = (fn: (r: BenchmarkRun) => number) => ({
			mean: mean(modelRuns.map(fn)),
			stddev: stddev(modelRuns.map(fn)),
		});
		summary[key] = {
			pass_rate: stat((r) => r.result.pass_rate),
			time_seconds: stat((r) => r.result.time_seconds),
			total_tokens: stat((r) => r.result.total_tokens),
			cost_usd: stat((r) => r.result.cost_usd),
		};
	}
	return summary;
}

function formatSuccessLine(
	completed: number,
	total: number,
	entry: RunEntry,
	result: BenchmarkRun,
): string {
	const r = result.result;
	const passColor = rateColor(r.pass_rate);
	return (
		`  ${pc.green("✓")} [${completed}/${total}] ` +
		`eval ${entry.evalCase.id} · ${entry.provider.modelId} · ` +
		`run ${entry.runNumber} · ` +
		`${passColor(`${(r.pass_rate * 100).toFixed(0)}%`)} ${pc.dim(`(${r.passed}/${r.total})`)} · ` +
		`${pc.dim(`${r.time_seconds.toFixed(1)}s · ${r.total_tokens} tok · $${r.cost_usd.toFixed(4)} · ${r.tool_calls} calls`)}`
	);
}

export async function runOrchestrator(
	config: ResolvedConfig,
	evals: EvalCase[],
	skillDir: string,
	systemPrompt: string,
	opts: {
		concurrency?: number;
		onProgress?: (msg: string) => void;
	} = {},
): Promise<OrchestratorResult> {
	const { onProgress } = opts;
	const log = onProgress ?? ((msg: string) => process.stdout.write(`${msg}\n`));

	const graderProvider = createProvider({
		name: config.grader.provider,
		model: config.grader.model,
		apiKey: config.grader.apiKey,
	});

	const matrix: RunEntry[] = [];
	for (const evalCase of evals) {
		for (const providerConfig of config.providers) {
			const provider = createProvider(providerConfig);
			for (let run = 1; run <= config.settings.runsPerProvider; run++) {
				matrix.push({ evalCase, provider, runNumber: run });
			}
		}
	}

	const concurrency = opts.concurrency ?? Math.min(matrix.length, MAX_CONCURRENCY);

	log(
		pc.bold(
			`Running ${matrix.length} eval(s): ${evals.length} eval(s) × ${config.providers.length} provider(s) × ${config.settings.runsPerProvider} run(s)`,
		),
	);

	const results: BenchmarkRun[] = [];
	const spinner = new Spinner();
	let completed = 0;

	async function executeOne(entry: RunEntry): Promise<BenchmarkRun> {
		const evalTimeoutMs = config.settings.timeout * 1000;
		const sandboxDir = fs.mkdtempSync(path.join(os.tmpdir(), "skill-eval-"));

		try {
			seedSandbox(sandboxDir, skillDir, entry.evalCase.files);
			const id = taskId(entry);

			return await withTimeout(
				(async () => {
					const turns =
						entry.evalCase.turns ?? (entry.evalCase.prompt ? [entry.evalCase.prompt] : []);
					const agentRun = await runAgentLoop({
						provider: entry.provider,
						system: systemPrompt,
						turns,
						tools: defaultTools,
						toolHandlers: createToolHandlers(sandboxDir, config.settings.timeout),
						maxSteps: config.settings.maxSteps,
						temperature: config.settings.temperature,
						onActivity: (detail) => spinner.detail(id, detail),
						checkTurnRelevance: turns.length > 1 ? createTurnChecker(graderProvider) : undefined,
					});

					const outputFiles = collectOutputFiles(sandboxDir);

					spinner.detail(id, "grading…");
					const grading: GradingResult = await gradeRun(
						graderProvider,
						entry.evalCase,
						agentRun,
						outputFiles,
					);

					const cost = estimateCost(
						entry.provider.modelId,
						agentRun.totalInputTokens,
						agentRun.totalOutputTokens,
					);

					return {
						eval_id: entry.evalCase.id,
						provider: entry.provider.name,
						model: entry.provider.modelId,
						run_number: entry.runNumber,
						result: {
							pass_rate: grading.pass_rate,
							passed: grading.passed,
							failed: grading.failed,
							total: grading.total,
							time_seconds: agentRun.totalLatencyMs / 1000,
							input_tokens: agentRun.totalInputTokens,
							output_tokens: agentRun.totalOutputTokens,
							total_tokens: agentRun.totalInputTokens + agentRun.totalOutputTokens,
							tool_calls: agentRun.totalToolCalls,
							errors: agentRun.errors,
							cost_usd: cost,
						},
						expectations: grading.expectations,
						claims: grading.claims,
						eval_feedback: grading.eval_feedback,
						error: null,
					};
				})(),
				evalTimeoutMs,
				`Eval timed out after ${config.settings.timeout}s`,
			);
		} finally {
			fs.rmSync(sandboxDir, { recursive: true, force: true });
		}
	}

	async function runTask(entry: RunEntry): Promise<void> {
		const id = taskId(entry);
		spinner.track(id, taskLabel(entry));

		let attempt = 0;
		while (true) {
			try {
				const result = await executeOne(entry);
				results.push(result);
				completed++;
				spinner.succeed(id, formatSuccessLine(completed, matrix.length, entry, result));
				return;
			} catch (err) {
				attempt++;
				const errMsg = extractErrorMessage(err);

				if (isRetryable(err) && attempt <= MAX_RETRIES) {
					const delay = RETRY_BASE_MS * 2 ** (attempt - 1);
					spinner.detail(
						id,
						`retrying in ${(delay / 1000).toFixed(0)}s (${attempt + 1}/${MAX_RETRIES + 1})`,
					);
					await sleep(delay);
					spinner.untrack(id);
					spinner.track(id, taskLabel(entry));
					continue;
				}

				results.push(buildFailureResult(entry, errMsg));
				completed++;
				spinner.succeed(
					id,
					`  ${pc.red("✗")} [${completed}/${matrix.length}] ` +
						`eval ${entry.evalCase.id} · ${entry.provider.modelId} · ` +
						`run ${entry.runNumber} · ${pc.red(errMsg)}`,
				);
				return;
			}
		}
	}

	spinner.start(matrix.length);

	const workers = Array.from({ length: Math.min(concurrency, matrix.length) }, async () => {
		while (matrix.length > 0) {
			const entry = matrix.shift();
			if (!entry) break;
			await runTask(entry);
		}
	});
	await Promise.all(workers);

	spinner.stop();

	return { runs: results, providerSummary: computeProviderSummary(results) };
}
